### PR TITLE
Groups API  + Tests + Example v2

### PR DESCRIPTION
**Pull request by abigalekim**

### CODE REVIEW

File: tiledb/api/examples/groups.rs
Comment (rroelke) [inline tiledb/api/examples/groups.rs:111]: Remove before merging

File: tiledb/api/src/context.rs
Comment (rroelke) [inline tiledb/api/src/context.rs:39]: Porting this comment from the other PR - this should parallel what you did for `Mode`

File: tiledb/api/src/error.rs
Comment (rroelke) [inline tiledb/api/src/error.rs:93]: In this case you could probably do `user_type: ObjectType`.
For datatype this field is something like `u32` or `f64` or what have you, so the mismatch is between the user programattic type and the TileDB data type, which isn't the same as the mismatch here

File: tiledb/api/src/error.rs
Comment (rroelke) [inline tiledb/api/src/error.rs:122]: Same as above

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:80]: This should be `std::convert::From`

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:115]: "create" sounds like some work is going on, this should probably just be called `new`

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:126]: I don't think we should enable any external APIs with `ffi`.  So this should probably just be `pub(crate)`.

`from_raw` should probably also be part of the name, since that's what this is doing.

This function should be marked `unsafe` since it is not safe on all possible inputs.

But overall I'm not sure you need this - there's just one call site right?  You could just put the `fn_typed` there calling `Metadata::new`

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:150]: Per my above comment I think this should be at most `pub(crate)`

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:551]: In my opinion it should not be possible to construct a `Group` if the group is *not* open.

I'm fine with this being present as a failsafe mechanism *somewhere* but:
1) I think it should probably  be in `open`, to prevent constructing the `Group` (unless there are APIs which are valid to call on a closed group? is that the case?); and/or
2) it should be remarked upon with a comment that we don't really expect it ever to find a closed group.

**Review by rroelke — CHANGES_REQUESTED** [review]
Seems like a lot of the comments I left on the previous PR still apply here.

File: tiledb/api/src/group/mod.rs
Comment (abigalekim) [reply tiledb/api/src/group/mod.rs:497]: Wait no, the Group is constructed in create, and then opened once it is created. And the function open itself returns a Self, so by construction it is open? And close is called upon dropping an object.

Comment (rroelke) [reply tiledb/api/src/group/mod.rs:497]: I should be clearer - a user of the API should not be able to get a Group type object unless the group it represents is open.  Your code currently does this, since `Group::open` returns `Err` if `tiledb_group_open` fails.  So the check here to `tiledb_group_is_open` in the destructor is kind of misleading, because it communicates that we might have a Group type object which represents a non-open group.

Unless the group can spontaneously close - I sure hope it can't!

I think you can add clarity to the use of `tiledb_group_is_open` by either calling it in `open` to prevent `open` from returning successfully, or by leaving a hard-to-miss comment here that we don't ever expect the group to be closed at this point.

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:148]: Remove any `println`s before merging

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:282]: Does this ever return `None`?

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:330]: Once you remove the `println`s I think you probably won't need the `clone` any more

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:444]: This C API doc doesn't mention anything about the group open mode but this sure sounds like a mutating function call...

Anyway, as I've been mulling over the whole "some methods are `&mut self` and some aren't" thing, it's kind of occurred to me that `&mut self` to represent a write mode method doesn't really work because the API allows you to just get another handle, and then you have ownership of that second handle and can call whatever you want anyway. And this can go in the direction of "just let the C API give back the error and don't worry about it" or in the direction of "embed the permissions to call certain APIs in the type system".  If we had infinite time I'd want the latter but let's move ahead with the former for now.

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:547]: I suppose it's something of a stylistic preference, but you can avoid this function call by using inner blocks instead.
```
{
    let group1_err = Group::open(&tdb, group1_uri.to_owned(), QueryType::Read)?;
    let res = group1_err.put_metadata(...);
    ...
    /* `res` will be dropped here, then `group1_err` */
}
```
I prefer this myself as it will drop everything used for this part of the test and be a better visual cue via indentation.

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:110]: I think this is probably the root cause of one of the bugs that we looked at, with the `f32` being a problem. This constructor doesn't require any correlation between `datatype` and `vec`.

I don't think we've really flushed out that story at all, but for now you can definitely catch *some* problems, like the one you encountered, by checking `std::mem::size_of::<T>()` against `datatype.size()` and making sure that they match.

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:721]: Is there an assert that can be used to express what this wants instead?
If it's hard to check the contents of the string then asserting none/some seems fine.

For the `dump` API to be useful anyway it probably need to be parsed to some useful representation. If it's not documented then all anyone can do with it is print it or check some/none as you are doing here.

File: tiledb/api/src/group/mod.rs
Comment (abigalekim) [reply tiledb/api/src/group/mod.rs:282]: Yes, when the dump is empty.

File: tiledb/api/src/error.rs
Comment (abigalekim) [reply tiledb/api/src/error.rs:92]: Oh wait, yea sorry! Actually this invalid discriminant error is used in the C API conversion, since the CAPI stores this enum as ints under the hood.

File: tiledb/api/src/group/mod.rs
Comment (abigalekim) [reply tiledb/api/src/group/mod.rs:493]: Fixed this!

File: tiledb/api/src/metadata.rs
Comment (abigalekim) [reply tiledb/api/src/metadata.rs:110]: Added this check, but keeping this comment since I think this may need to be an additional story/discussed.

File: tiledb/api/src/error.rs
Comment (rroelke) [reply tiledb/api/src/error.rs:92]: This comment is in reference to the `TypeMismatch` variant. I don't follow your comment.

Is `TypeMismatch` used anywhere? I checked out your branch and deleted it and it doesn't appear to be.  Better to remove it.  Looks the same for `ModeErrorKind`.

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [reply tiledb/api/src/group/mod.rs:508]: Oh, good point, you have a consuming method `delete_group` and then also this `Drop` impl.

Please leave a comment here though because it is not obvious why this is needed

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [reply tiledb/api/src/group/mod.rs:295]: Which LOC does that? I don't see anything returning `None`

File: tiledb/api/src/metadata.rs
Comment (rroelke) [inline tiledb/api/src/metadata.rs:121]: This should be `DataTypeErrorKind::TypeMismatch`

File: tiledb/api/src/metadata.rs
Comment (davisp) [reply tiledb/api/src/metadata.rs]: `datatype.is_compatible_type::<T>() -> bool` is probably what you want here.

File: tiledb/api/src/group/mod.rs
Comment (abigalekim) [reply tiledb/api/src/group/mod.rs]: Fixed.

File: tiledb/api/src/group/mod.rs
Comment (abigalekim) [reply tiledb/api/src/group/mod.rs:510]: Wrote a comment!

File: tiledb/api/src/error.rs
Comment (abigalekim) [reply tiledb/api/src/error.rs]: Deleted the type mismatch errors.

File: tiledb/api/src/metadata.rs
Comment (davisp) [inline tiledb/api/src/metadata.rs:21]: Lets remove commented out code here.

**Review by davisp — COMMENTED** [review]
Skimming this, I'm not seeing anything of concern. Just the one minor nit. I'll defer to @rroelke to approve since he's reviewed this more than me.

File: tiledb/api/src/group/mod.rs
Comment (rroelke) [inline tiledb/api/src/group/mod.rs:107]: `expect` isn't appropriate here since we can return `Result`

also I think you can remove this if you want, your comment yesterday (earlier this week? something) in `Drop` convinced me that you did it correctly the first time

**Review by rroelke — APPROVED** [review]
All righty!
Looks like there are two final comments but they should be pretty trivial to address.
Let's get this in there, nice work!

**Pull request merged**
